src/track/serato/beatgrid.cpp

```diff
@@ -0,0 +1,401 @@
+        return false;
+    }
+
+    if (pTerminalMarker->bpm() <= 0) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "Terminal marker's BPM"
+                          << pTerminalMarker->bpm()
+                          << "must be greater than 0";
+        return false;
+    }
+
+    if (pTerminalMarker->positionSecs() < 0) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "Non-terminal marker has invalid position"
+                          << pTerminalMarker->positionSecs()
+                          << "< 0";
+        return false;
+    }
+
+    if (pTerminalMarker->positionSecs() <= previousBeatPositionMillis) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "Terminal marker's position"
+                          << pTerminalMarker->positionSecs()
+                          << "must be greater than the previous marker's position"
+                          << previousBeatPositionMillis;
+        return false;
+    }
+
+    // Read footer
+    //
+    // FIXME: This byte has caused some headache because I have not the
+    // slightest idea what this value could be. Apparently it's random, because
+    // it changes even when entering Serato's "Edit Grid" mode and then leaving
+    // it immediately without making any changes.
+    // For now, we only read it to be able to dump the exact same byte sequence
+    // later on.
+    quint8 footer;
+    stream >> footer;
+
+    if (stream.status() != QDataStream::Status::Ok) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "Stream read failed with status" << stream.status();
+        return false;
+    }
+
+    if (!stream.atEnd()) {
+        kLogger.warning() << "Parsing SeratoBeatGrid failed:"
+                          << "Unexpected trailing data";
+        return false;
+    }
+    seratoBeatGrid->setNonTerminalMarkers(std::move(nonTerminalMarkers));
+    seratoBeatGrid->setTerminalMarker(pTerminalMarker);
+    seratoBeatGrid->setFooter(footer);
+
+    return true;
+}
+
+bool SeratoBeatGrid::parseBase64Encoded(
+        SeratoBeatGrid* seratoBeatGrid, const QByteArray& base64EncodedData) {
+    const auto decodedData = QByteArray::fromBase64(base64EncodedData);
+    if (!decodedData.startsWith(kSeratoBeatGridBase64EncodedPrefix)) {
+        kLogger.warning() << "Decoding SeratoBeatGrid from base64 failed:"
+                          << "Unexpected prefix";
+        return false;
+    }
+    DEBUG_ASSERT(decodedData.size() >= kSeratoBeatGridBase64EncodedPrefix.size());
+    if (!parseID3(
+                seratoBeatGrid,
+                decodedData.mid(kSeratoBeatGridBase64EncodedPrefix.size()))) {
+        kLogger.warning() << "Parsing base64encoded SeratoBeatGrid failed!";
+        return false;
+    }
+
+    return true;
+}
+
+QByteArray SeratoBeatGrid::dump(taglib::FileType fileType) const {
```

Comment:
dump to what? Does it make sense to immediately create the TagLib::String()? 
Than the function can be named dumpToTagLibString() 

Note: TagLib::String() uses internally a TagLib::wstring just like QString. 

Not sure if we can make use of it, by configuring the QDataStream to generate wchars directly ... mm

All these considerations are probably not worth the time :-/

Reply:
> dump to what? Does it make sense to immediately create the TagLib::String()?

No, I think makes sense to dump it to a plain QByteArray. Depending on the tag type this can be written to a `TagLib::String` or a `TagLib::ByteVector`. Also, it makes testing easier and if if we ever switch to another tag library we do not need to touch this. All taglib-related code remains in `src/track/taglib`.

Reply:
OK. 
I was originally really struggling to what the value is dumped.  
Can we find a better function name or add a comment?

Reply:
I'll add a comment.